### DISCUSSION TITLE
[A-L] (19 febbraio 2024) Esercizio 3

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#85] [Esami A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/821e8207-c2a7-4b0b-b364-d45b37528ed7)

**Comment by Frisbon** [comment]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/1d0d46f4-eece-4d63-abd6-985bf919c865)
Scorro l'array dei Padri in O(n) e mi salvo chi compare (quindi chi non è foglia)
Poi, basta che scorro su VIS in O(n) e aggiungo alla soluzione gli indici che hanno VIS impostato a 0

**Comment by IonutCicio** [comment]
```rust
fn leaves(daddies: &[usize]) -> Vec<usize> {
    let mut is_daddy = vec![false; daddies.len()];

    for daddy in 0..daddies.len() {
        is_daddy[daddies[daddy]] = true;
    }

    (0..daddies.len()).filter(|&x| !is_daddy[x]).collect()
}
```

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/264da91c-0d9c-438a-ab98-fff61e920d44)

**Comment by cek753** [comment]
```
alg(P):
    A = array di lunghezza len(P) inizializzato a 1
    sol = insieme vuoto
    for i in range(1,len(P)):
        A[P[i]]=0

    for i in range(1, len(A)):
        if A[i] == 1:
            sol.add(i)
return sol
```

**Comment by LAMASE** [comment]
La definizione nell'esercizio parla di "un solo vicino" cosa che porebbe accadere anche con la radice (albero di due soli nodi, catena ecc). 
Se va tenuta in considerazione questa possibilità basta tenere un contatore per ogni nodo (un arco conta sia per il nodo entrante che in quello uscente, escludendo la radice che nel vettore dei padri punta a se stesso) e tirar fuori i nodi con count == 1.

```
algo(P)
    sol = {}
    count = [0,0,....]
    for i in P
        if (P[i] != i) // se è radice non c'è un vero arco
            count[i]++
            count[P[i]]++
    for i in P
        if (count[i] == 1) sol = sol.add(i);
    return sol
```
